FP fix and documentation for `branches_sharing_code` lint

xFrednet opened the pull request:
Closes rust-lang/rust-clippy#7369

Related rust-lang/rust-clippy#7452 I'm still thinking about the best way to fix this. I could simply add another visitor to ensure that the moved expressions don't modify values being used in the condition, but I'm not totally happy with this due to the complexity. I therefore only documented it for now 


changelog: [`branches_sharing_code`] fixed false positive where block expressions would sometimes be ignored.

rust-highfive commented:
r? @camsteffen

(rust-highfive has picked a reviewer for you, use r? to override)

xFrednet commented on tests/ui/branches_sharing_code/false_positives.rs:
```diff
@@ -0,0 +1,48 @@
+    fn baz(&mut self) {}
+}
+
+// #########################
+// # Issue 7452
+// #########################
+fn test() {
+    let v = vec![1, 2, 3, 4, 5, 6, 7, 8, 9, 0];
+    let mut counter = 0;
+    for i in v {
+        if counter == 0 {
+            counter += 1;
+            println!("first");
+        } else if counter == 1 {
+            counter += 1;
+            println!("second");
+        } else {
+            counter += 1;
+            println!("other: {}", i);
+        }
+    }
+}
```
This is the issue that is not fixed yet. I found it reasonable to add a test for it (even if it fails) as it's also documented in the lint description

camsteffen replied:
I'm not sure we gain anything by adding this. Tests are valuable when they verify correct behavior.

xFrednet replied:
Yeah, I wasn't sure with this one that's, the reason why I pointed it out. I'll remove it from this PR and then squash the two commits :upside_down_face:

camsteffen commented on clippy_lints/src/copies.rs:
```diff
@@ -225,7 +228,7 @@ fn lint_same_then_else<'tcx>(
             expr,
         );
-    } else if end_eq != 0 || (has_expr && expr_eq) {
+    } else if (end_eq != 0 && !has_expr) || (has_expr && expr_eq) {
```
I think the real problem is that `scan_block_for_eq` can return `end_eq > 0 && expr_eq == false`? End statements should not be evaluated if the expressions are not equal.

xFrednet replied:
You're right. `scan_block_for_eq` even contained an if block for this purpose, but the condition was not quite correct, it's fixed in the latest version. :upside_down_face:

camsteffen commented:
For #7452, I would maybe first collect all the variables referenced in the condition. Then, when determining if statements are equal, if there is a mutable reference to the variable, the statements are not equal. So I think you would add `SpanlessEq::disallowed_mut_locals: FxHashSet<HirId>`.

xFrednet commented:
> For #7452, I would maybe first collect all the variables referenced in the condition. Then, when determining if statements are equal, if there is a mutable reference to the variable, the statements are not equal. So I think you would add `SpanlessEq::disallowed_mut_locals: FxHashSet<HirId>`.

Interesting idea, I thought about expending the `UsedValueFinderVisitor` to also collect mutations. Putting it into `SpanlessEq` feels like the wrong place, as it's not really part of the comparison in my eyes, and I feel like `SpanlessEq` is already quite complex :thinking:

xFrednet commented:
The FP is now fixed in `scan_block_for_eq` where it originated from (very nice catch!!). I've squashed the commits right away, having a fix up commit that just reverted the previous implementation and the FP test seemed a bit unnecessary. I hope that is okay for you :upside_down_face:

camsteffen commented:
> Interesting idea, I thought about expending the `UsedValueFinderVisitor` to also collect mutations. Putting it into `SpanlessEq` feels like the wrong place, as it's not really part of the comparison in my eyes, and I feel like `SpanlessEq` is already quite complex 🤔

Yeah I know what you mean. That would work. One more idea for `SpanlessEq` - `expr_fallback` could have a counterpart `expr_precondition`.

camsteffen commented:
Looks good, thanks!

@bors r+

bors commented:
:pushpin: Commit 61e280863fb68c26bb1f976acfa4a45a26d70545 has been approved by `camsteffen`

<!-- @bors r=camsteffen 61e280863fb68c26bb1f976acfa4a45a26d70545 -->
<!-- homu: {"type":"Approved","sha":"61e280863fb68c26bb1f976acfa4a45a26d70545","approver":"camsteffen"} -->

bors commented:
:hourglass: Testing commit 61e280863fb68c26bb1f976acfa4a45a26d70545 with merge 2b193e247f108c303d2ef9818ddd4aadceaf5760...
<!-- homu: {"type":"BuildStarted","head_sha":"61e280863fb68c26bb1f976acfa4a45a26d70545","merge_sha":"2b193e247f108c303d2ef9818ddd4aadceaf5760"} -->

xFrednet commented:
> Yeah I know what you mean. That would work. One more idea for `SpanlessEq` - `expr_fallback` could have a counterpart `expr_precondition`.

`SpanlessEq` - `expr_fallback` doesn't ring a bell for me right now, but it also has been some time since I’ve dug through the code :thinking: Thank you for the suggestions :)

bors commented:
:sunny: Test successful - [checks-action_dev_test](https://github.com/rust-lang/rust-clippy/runs/3070671145), [checks-action_remark_test](https://github.com/rust-lang/rust-clippy/runs/3070665394), [checks-action_test](https://github.com/rust-lang/rust-clippy/runs/3070763622)
Approved by: camsteffen
Pushing 2b193e247f108c303d2ef9818ddd4aadceaf5760 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"camsteffen","base_ref":"master","builders":{"checks-action_test":"https://github.com/rust-lang/rust-clippy/runs/3070763622","checks-action_dev_test":"https://github.com/rust-lang/rust-clippy/runs/3070671145","checks-action_remark_test":"https://github.com/rust-lang/rust-clippy/runs/3070665394"},"merge_sha":"2b193e247f108c303d2ef9818ddd4aadceaf5760"} -->
